File: AugmentedRealityWithArucoMarkers/augmented_reality_with_aruco.cpp
// Original example by Sunita Nayak at BigVision LLC, based on OpenCV.
// Modernized for current OpenCV APIs and reproducible headless execution.

#include <opencv2/core/utility.hpp>
#include <opencv2/core/version.hpp>
#if CV_VERSION_MAJOR >= 5
#include <opencv2/geometry/2d.hpp>
#endif
#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/objdetect/aruco_detector.hpp>
#include <opencv2/objdetect/aruco_dictionary.hpp>
#include <opencv2/videoio.hpp>

#include <algorithm>
#include <cmath>
#include <filesystem>
#include <iostream>
#include <optional>
#include <string>
#include <vector>

namespace fs = std::filesystem;

namespace {

const char* kKeys =
    "{help h usage ? |      | Show this message }"
    "{image i        |      | Input image; defaults to test.jpg }"
    "{video v        |      | Input video }"
    "{camera c       |-1    | Camera device index }"
    "{overlay        |new_scenery.jpg| Image placed in the marker region }"
    "{output o       |      | Output image or AVI path }"
    "{augmented-only |false | Write only the augmented view }"
    "{strict         |false | Fail if no frame can be augmented }"
    "{display        |false | Show a GUI window }";

struct AugmentResult {
    cv::Mat image;
    bool augmented = false;
    std::vector<int> detected_ids;
};

std::optional<std::size_t> markerIndex(
    const std::vector<int>& ids,
    int marker_id
) {
    const auto iterator = std::find(ids.begin(), ids.end(), marker_id);
    if (iterator == ids.end()) {
        return std::nullopt;
    }
    return static_cast<std::size_t>(std::distance(ids.begin(), iterator));
}

std::optional<std::vector<cv::Point2f>> destinationPoints(
    const std::vector<std::vector<cv::Point2f>>& marker_corners,
    const std::vector<int>& marker_ids,
    float margin_fraction = 0.02F
) {
    const auto top_left_index = markerIndex(marker_ids, 25);
    const auto top_right_index = markerIndex(marker_ids, 33);
    const auto bottom_right_index = markerIndex(marker_ids, 30);
    const auto bottom_left_index = markerIndex(marker_ids, 23);
    if (!top_left_index || !top_right_index ||
        !bottom_right_index || !bottom_left_index) {
        return std::nullopt;
    }

    const cv::Point2f top_left = marker_corners.at(*top_left_index).at(1);
    const cv::Point2f top_right = marker_corners.at(*top_right_index).at(2);
    const cv::Point2f bottom_right =
        marker_corners.at(*bottom_right_index).at(0);
    const cv::Point2f bottom_left = marker_corners.at(*bottom_left_index).at(0);
    const float top_edge_length = cv::norm(top_left - top_right);
    if (top_edge_length <= 1.0F) {
        throw std::runtime_error(
            "The top marker corners are too close to define a region."
        );
    }
    const float margin = margin_fraction * top_edge_length;
    return std::vector<cv::Point2f>{
        {top_left.x - margin, top_left.y - margin},
        {top_right.x + margin, top_right.y - margin},
        {bottom_right.x + margin, bottom_right.y + margin},
        {bottom_left.x - margin, bottom_left.y + margin},
    };
}

AugmentResult augmentFrame(
    const cv::Mat& frame,
    const cv::Mat& overlay,
    const cv::aruco::ArucoDetector& detector
) {
    if (frame.empty()) {
        throw std::invalid_argument("The input frame is empty.");
    }
    if (overlay.empty()) {
        throw std::invalid_argument("The overlay image is empty.");
    }

    std::vector<std::vector<cv::Point2f>> marker_corners;
    std::vector<std::vector<cv::Point2f>> rejected_candidates;
    std::vector<int> marker_ids;
    detector.detectMarkers(
        frame, marker_corners, marker_ids, rejected_candidates
    );

    AugmentResult result{frame.clone(), false, marker_ids};
    std::sort(result.detected_ids.begin(), result.detected_ids.end());
    const auto points_destination =
        destinationPoints(marker_corners, marker_ids);
    if (!points_destination) {
        return result;
    }

    const std::vector<cv::Point2f> points_source{
        {0.0F, 0.0F},
        {static_cast<float>(overlay.cols - 1), 0.0F},
        {
            static_cast<float>(overlay.cols - 1),
            static_cast<float>(overlay.rows - 1)
        },
        {0.0F, static_cast<float>(overlay.rows - 1)},
    };
    const cv::Mat transform =
        cv::getPerspectiveTransform(points_source, *points_destination);
    cv::Mat warped;
    cv::warpPerspective(
        overlay, warped, transform, frame.size(), cv::INTER_CUBIC
    );

    std::vector<cv::Point> polygon;
    polygon.reserve(points_destination->size());
    for (const cv::Point2f& point : *points_destination) {
        polygon.emplace_back(cvRound(point.x), cvRound(point.y));
    }
    cv::Mat mask = cv::Mat::zeros(frame.size(), CV_8UC1);
    cv::fillConvexPoly(mask, polygon, cv::Scalar(255), cv::LINE_AA);
    const cv::Mat erosion_element =
        cv::getStructuringElement(cv::MORPH_RECT, cv::Size(3, 3));
    cv::erode(mask, mask, erosion_element, cv::Point(-1, -1), 3);

    warped.copyTo(result.image, mask);
    result.augmented = true;
    return result;
}

cv::Mat composeOutput(
    const cv::Mat& original,
    const cv::Mat& augmented,
    bool augmented_only
) {
    if (augmented_only) {
        return augmented;
    }
    cv::Mat side_by_side;
    cv::hconcat(original, augmented, side_by_side);
    return side_by_side;
}

fs::path imageOutputPath(const fs::path& input_path) {
    const std::string extension =
        input_path.has_extension() ? input_path.extension().string() : ".jpg";
    return input_path.parent_path() /
           fs::path(input_path.stem().string() + "_ar_out_cpp" + extension);
}

fs::path videoOutputPath(const fs::path& input_path) {
    if (input_path.empty()) {
        return "ar_out_cpp.avi";
    }
    return input_path.parent_path() /
           fs::path(input_path.stem().string() + "_ar_out_cpp.avi");
}

void createParentDirectory(const fs::path& output_path) {
    if (output_path.has_parent_path()) {
        fs::create_directories(output_path.parent_path());
    }
}

void printDetectedIds(const std::vector<int>& ids) {
    if (ids.empty()) {
        std::cout << "none";
        return;
    }
    for (std::size_t index = 0; index < ids.size(); ++index) {
        if (index > 0) {
            std::cout << ',';
        }
        std::cout << ids[index];
    }
}

}  // namespace

int main(int argc, char** argv) {
    cv::CommandLineParser parser(argc, argv, kKeys);
    parser.about("Augmented reality with four ArUco markers");
    if (parser.has("help")) {
        parser.printMessage();
        return 0;
    }
    if (!parser.check()) {
        parser.printErrors();
        return 2;
    }

    std::string image_path = parser.get<std::string>("image");
    const std::string video_path = parser.get<std::string>("video");
    const int camera = parser.get<int>("camera");
    const std::string overlay_path = parser.get<std::string>("overlay");
    const std::string requested_output = parser.get<std::string>("output");
    const bool augmented_only = parser.get<bool>("augmented-only");
    const bool strict = parser.get<bool>("strict");
    const bool display = parser.get<bool>("display");

    const int input_count =
        static_cast<int>(!image_path.empty()) +
        static_cast<int>(!video_path.empty()) +
        static_cast<int>(camera >= 0);
    if (input_count > 1) {
        std::cerr << "Error: choose only one of --image, --video, or --camera.\n";
        return 2;
    }
    if (input_count == 0) {
        image_path = "test.jpg";
    }

    const cv::Mat overlay = cv::imread(overlay_path, cv::IMREAD_COLOR);
    if (overlay.empty()) {
        std::cerr << "Error: overlay image not found or unreadable: "
                  << overlay_path << '\n';
        return 2;
    }

    const cv::aruco::Dictionary dictionary =
        cv::aruco::getPredefinedDictionary(cv::aruco::DICT_6X6_250);
    const cv::aruco::DetectorParameters parameters;
    const cv::aruco::ArucoDetector detector(dictionary, parameters);

    try {
        if (!image_path.empty()) {
            const cv::Mat frame =
                cv::imread(image_path, cv::IMREAD_COLOR);
            if (frame.empty()) {
                std::cerr << "Error: input image not found or unreadable: "
                          << image_path << '\n';
                return 2;
            }
            const AugmentResult result =
                augmentFrame(frame, overlay, detector);
            if (strict && !result.augmented) {
                std::cerr << "Error: required marker IDs 25, 33, 30, and 23 "
                             "were not all detected.\n";
                return 3;
            }
            const cv::Mat output =
                composeOutput(frame, result.image, augmented_only);
            const fs::path output_path = requested_output.empty()
                ? imageOutputPath(image_path)
                : fs::path(requested_output);
            createParentDirectory(output_path);
            if (!cv::imwrite(output_path.string(), output)) {
                std::cerr << "Error: OpenCV could not write: "
                          << output_path << '\n';
                return 3;
            }
            std::cout << "Wrote " << output_path << ": "
                      << (result.augmented ? "augmented" : "unchanged")
                      << "; detected IDs ";
            printDetectedIds(result.detected_ids);
            std::cout << '\n';

            if (display) {
                cv::imshow("AR using ArUco markers", output);
                cv::waitKey(0);
                cv::destroyAllWindows();
            }
            return 0;
        }

        cv::VideoCapture capture;
        if (!video_path.empty()) {
            capture.open(video_path);
        } else {
            capture.open(camera);
        }
        if (!capture.isOpened()) {
            std::cerr << "Error: could not open "
                      << (!video_path.empty() ? video_path : "camera") << ".\n";
            return 2;
        }

        const fs::path output_path = requested_output.empty()
            ? videoOutputPath(video_path)
            : fs::path(requested_output);
        createParentDirectory(output_path);
        double fps = capture.get(cv::CAP_PROP_FPS);
        if (!std::isfinite(fps) || fps <= 0.0) {
            fps = 28.0;
        }

        cv::VideoWriter writer;
        int frame_count = 0;
        int augmented_count = 0;
        while (true) {
            cv::Mat frame;
            if (!capture.read(frame)) {
                break;
            }
            const AugmentResult result =
                augmentFrame(frame, overlay, detector);
            const cv::Mat output =
                composeOutput(frame, result.image, augmented_only);
            if (!writer.isOpened()) {
                writer.open(
                    output_path.string(),
                    cv::VideoWriter::fourcc('M', 'J', 'P', 'G'),
                    fps,
                    output.size()
                );
                if (!writer.isOpened()) {
                    std::cerr << "Error: could not create video: "
                              << output_path << '\n';
                    return 3;
                }
            }
            writer.write(output);
            ++frame_count;
            augmented_count += static_cast<int>(result.augmented);

            if (display) {
                cv::imshow("AR using ArUco markers", output);
                const int key = cv::waitKey(1) & 0xFF;
                if (key == 27 || key == 'q') {
                    break;
                }
            }
        }
        capture.release();
        writer.release();
        if (display) {
            cv::destroyAllWindows();
        }

        if (frame_count == 0) {
            std::cerr << "Error: the input produced no readable frames.\n";
            return 3;
        }
        if (strict && augmented_count == 0) {
            std::cerr << "Error: no frame contained all required markers.\n";
            return 3;
        }
        std::cout << "Wrote " << output_path << ": augmented "
                  << augmented_count << '/' << frame_count << " frames\n";
    } catch (const cv::Exception& error) {
        std::cerr << "OpenCV error: " << error.what() << '\n';
        return 3;
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        return 3;
    }
    return 0;
}

File: AugmentedRealityWithArucoMarkers/augmented_reality_with_aruco.py
#!/usr/bin/env python3
"""Replace a four-marker quadrilateral with an image using current ArUco APIs."""

# Original example by Sunita Nayak at BigVision LLC, based on OpenCV.
# Modernized for current OpenCV APIs and reproducible headless execution.

from __future__ import annotations

import argparse
from pathlib import Path
import sys

import cv2 as cv
import numpy as np


PROJECT_DIR = Path(__file__).resolve().parent
DEFAULT_IMAGE = PROJECT_DIR / "test.jpg"
DEFAULT_OVERLAY = PROJECT_DIR / "new_scenery.jpg"
REQUIRED_MARKER_CORNERS = ((25, 1), (33, 2), (30, 0), (23, 0))


def create_detector():
    """Create the modern stateful ArUco detector once for all frames."""
    if not hasattr(cv, "aruco") or not hasattr(cv.aruco, "ArucoDetector"):
        raise RuntimeError(
            "This OpenCV build does not provide cv.aruco.ArucoDetector. "
            "Install OpenCV 4.8 or newer."
        )
    dictionary = cv.aruco.getPredefinedDictionary(cv.aruco.DICT_6X6_250)
    parameters = cv.aruco.DetectorParameters()
    return cv.aruco.ArucoDetector(dictionary, parameters)


def destination_points(
    marker_corners,
    marker_ids,
    margin_fraction: float = 0.02,
) -> np.ndarray | None:
    """Return the four destination corners, or None if a marker is absent."""
    if marker_ids is None:
        return None
    if margin_fraction < 0:
        raise ValueError("margin_fraction must not be negative.")

    ids = np.asarray(marker_ids).reshape(-1)
    corners_by_id = {
        int(marker_id): np.asarray(marker_corners[index]).reshape(4, 2)
        for index, marker_id in enumerate(ids)
    }
    if any(marker_id not in corners_by_id for marker_id, _ in REQUIRED_MARKER_CORNERS):
        return None

    references = [
        corners_by_id[marker_id][corner_index]
        for marker_id, corner_index in REQUIRED_MARKER_CORNERS
    ]
    top_left, top_right, bottom_right, bottom_left = references
    top_edge_length = float(np.linalg.norm(top_left - top_right))
    if top_edge_length <= 1.0:
        raise ValueError("The top marker corners are too close to define a region.")
    margin = margin_fraction * top_edge_length

    return np.asarray(
        [
            [top_left[0] - margin, top_left[1] - margin],
            [top_right[0] + margin, top_right[1] - margin],
            [bottom_right[0] + margin, bottom_right[1] + margin],
            [bottom_left[0] - margin, bottom_left[1] + margin],
        ],
        dtype=np.float32,
    )


def augment_frame(
    frame: np.ndarray,
    overlay: np.ndarray,
    detector=None,
    margin_fraction: float = 0.02,
) -> tuple[np.ndarray, bool, list[int]]:
    """Warp overlay into the marker frame and preserve input when markers are missing."""
    if frame is None or frame.size == 0:
        raise ValueError("The input frame is empty.")
    if overlay is None or overlay.size == 0:
        raise ValueError("The overlay image is empty.")
    if detector is None:
        detector = create_detector()

    marker_corners, marker_ids, _ = detector.detectMarkers(frame)
    detected_ids = (
        [] if marker_ids is None else sorted(int(value) for value in marker_ids.reshape(-1))
    )
    points_destination = destination_points(
        marker_corners, marker_ids, margin_fraction
    )
    if points_destination is None:
        return frame.copy(), False, detected_ids

    height, width = overlay.shape[:2]
    points_source = np.asarray(
        [[0, 0], [width - 1, 0], [width - 1, height - 1], [0, height - 1]],
        dtype=np.float32,
    )
    transform = cv.getPerspectiveTransform(points_source, points_destination)
    warped = cv.warpPerspective(
        overlay,
        transform,
        (frame.shape[1], frame.shape[0]),
        flags=cv.INTER_CUBIC,
    )

    mask = np.zeros(frame.shape[:2], dtype=np.uint8)
    polygon = np.rint(points_destination).astype(np.int32)
    cv.fillConvexPoly(mask, polygon, 255, lineType=cv.LINE_AA)
    erosion_element = cv.getStructuringElement(cv.MORPH_RECT, (3, 3))
    mask = cv.erode(mask, erosion_element, iterations=3)

    augmented = frame.copy()
    augmented[mask > 0] = warped[mask > 0]
    return augmented, True, detected_ids


def compose_output(
    original: np.ndarray,
    augmented: np.ndarray,
    augmented_only: bool,
) -> np.ndarray:
    return augmented if augmented_only else cv.hconcat([original, augmented])


def default_image_output(input_path: Path) -> Path:
    suffix = input_path.suffix if input_path.suffix else ".jpg"
    return input_path.with_name(f"{input_path.stem}_ar_out_py{suffix}")


def default_video_output(input_path: Path | None) -> Path:
    if input_path is None:
        return PROJECT_DIR / "ar_out_py.avi"
    return input_path.with_name(f"{input_path.stem}_ar_out_py.avi")


def process_image(
    input_path: Path,
    overlay: np.ndarray,
    detector,
    output_path: Path,
    augmented_only: bool = False,
    strict: bool = False,
    display: bool = False,
) -> tuple[bool, list[int]]:
    frame = cv.imread(str(input_path), cv.IMREAD_COLOR)
    if frame is None:
        raise FileNotFoundError(f"Input image not found or unreadable: {input_path}")

    augmented, did_augment, detected_ids = augment_frame(frame, overlay, detector)
    if strict and not did_augment:
        raise RuntimeError(
            "Required marker IDs 25, 33, 30, and 23 were not all detected. "
            f"Detected: {detected_ids or 'none'}."
        )
    output = compose_output(frame, augmented, augmented_only)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    if not cv.imwrite(str(output_path), output):
        raise OSError(f"OpenCV could not write: {output_path}")

    if display:
        cv.imshow("AR using ArUco markers", output)
        cv.waitKey(0)
        cv.destroyAllWindows()
    return did_augment, detected_ids


def process_video(
    video_path: Path | None,
    camera: int | None,
    overlay: np.ndarray,
    detector,
    output_path: Path,
    augmented_only: bool = False,
    strict: bool = False,
    display: bool = False,
) -> tuple[int, int]:
    capture = cv.VideoCapture(camera if camera is not None else str(video_path))
    if not capture.isOpened():
        source = f"camera {camera}" if camera is not None else str(video_path)
        raise RuntimeError(f"Could not open {source}.")

    fps = capture.get(cv.CAP_PROP_FPS)
    if not np.isfinite(fps) or fps <= 0:
        fps = 28.0
    output_path.parent.mkdir(parents=True, exist_ok=True)

    writer = None
    frame_count = 0
    augmented_count = 0
    try:
        while True:
            has_frame, frame = capture.read()
            if not has_frame:
                break
            augmented, did_augment, _ = augment_frame(frame, overlay, detector)
            output = compose_output(frame, augmented, augmented_only)
            if writer is None:
                writer = cv.VideoWriter(
                    str(output_path),
                    cv.VideoWriter_fourcc(*"MJPG"),
                    fps,
                    (output.shape[1], output.shape[0]),
                )
                if not writer.isOpened():
                    raise RuntimeError(f"Could not create video: {output_path}")
            writer.write(output)
            frame_count += 1
            augmented_count += int(did_augment)

            if display:
                cv.imshow("AR using ArUco markers", output)
                key = cv.waitKey(1) & 0xFF
                if key in (27, ord("q")):
                    break
    finally:
        capture.release()
        if writer is not None:
            writer.release()
        if display:
            cv.destroyAllWindows()

    if frame_count == 0:
        raise RuntimeError("The input produced no readable frames.")
    if strict and augmented_count == 0:
        raise RuntimeError(
            "None of the frames contained all required marker IDs "
            "25, 33, 30, and 23."
        )
    return frame_count, augmented_count


def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Replace a four-ArUco-marker region with an image."
    )
    source = parser.add_mutually_exclusive_group()
    source.add_argument("--image", type=Path, help="Input image.")
    source.add_argument("--video", type=Path, help="Input video.")
    source.add_argument("--camera", type=int, help="Camera device index.")
    parser.add_argument(
        "--overlay",
        type=Path,
        default=DEFAULT_OVERLAY,
        help="Image to place inside the marker region.",
    )
    parser.add_argument("--output", type=Path, help="Output image or AVI path.")
    parser.add_argument(
        "--augmented-only",
        action="store_true",
        help="Write only the augmented view instead of a side-by-side comparison.",
    )
    parser.add_argument(
        "--strict",
        action="store_true",
        help="Fail if the required marker set cannot be augmented.",
    )
    parser.add_argument(
        "--display",
        action="store_true",
        help="Show results in a GUI window; press Q or Escape to stop video.",
    )
    return parser.parse_args(argv)


def main(argv: list[str] | None = None) -> int:
    args = parse_args(argv)
    input_image = args.image
    if input_image is None and args.video is None and args.camera is None:
        input_image = DEFAULT_IMAGE

    try:
        overlay = cv.imread(str(args.overlay), cv.IMREAD_COLOR)
        if overlay is None:
            raise FileNotFoundError(
                f"Overlay image not found or unreadable: {args.overlay}"
            )
        detector = create_detector()

        if input_image is not None:
            output_path = args.output or default_image_output(input_image)
            did_augment, detected_ids = process_image(
                input_image,
                overlay,
                detector,
                output_path,
                augmented_only=args.augmented_only,
                strict=args.strict,
                display=args.display,
            )
            status = "augmented" if did_augment else "unchanged (markers missing)"
            print(
                f"Wrote {output_path}: {status}; "
                f"detected IDs {detected_ids or 'none'}"
            )
        else:
            output_path = args.output or default_video_output(args.video)
            frame_count, augmented_count = process_video(
                args.video,
                args.camera,
                overlay,
                detector,
                output_path,
                augmented_only=args.augmented_only,
                strict=args.strict,
                display=args.display,
            )
            print(
                f"Wrote {output_path}: augmented {augmented_count}/"
                f"{frame_count} frames"
            )
    except (FileNotFoundError, OSError, RuntimeError, ValueError, cv.error) as error:
        print(f"Error: {error}", file=sys.stderr)
        return 2
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
